### DISCUSSION TITLE
IRAM is almost overflow. Is it okay ?

### DISCUSSION THREAD

**Discussion by sir-user** [micropython/micropython#12793] [Core Development]
`idf.py size` command console return :
```
Total sizes:
Used static DRAM:   53516 bytes (  71064 remain, 43.0% used)
      .data size:   21172 bytes
      .bss  size:   32344 bytes
Used static IRAM:  127514 bytes (   3558 remain, 97.3% used)
      .text size:  126487 bytes
   .vectors size:    1027 bytes
Used Flash size : 1546437 bytes
      .text     : 1218087 bytes
      .rodata   :  328094 bytes
Total image size: 1695123 bytes (.bin may be padded larger)
```

My thread function does not work right and I see my micropython allocate to CPU1 when I push reset button.

Could someone help me?

**Comment by jimmo** [comment]
Does your thread function use the `@micropython.native` decorator? At runtime that's the only thing that uses IRAM.

If you want to reduce IRAM usage in your compiled firmware, see the `MICROPY_WRAP_MP_` macros in `ports/esp32/mpconfigport.h`. If you comment out any of these then those corresponding functions will be placed in ROM instead.

**Reply by sir-user** [reply]
Hi jimmo, could you give me some more details? 
When I comment out to place `MICROPY_WRAP_MP_ ` macros in ROM and to build it by `idf.py build` command , I got a lots of issues on console and it does not built.
What do you mean "thread function"? I use built-in thread, it is `import _thread`
```
from libs import WavP
import _thread
from time import sleep_ms

Plyr = WavP.WavP()

_thread.start_new_thread(Plyr.play_wav,("wav_files/at.wav",))

sleep_ms(500)

Plyr.status = 0
```

**Reply by jimmo** [reply]
> When I comment out to place `MICROPY_WRAP_MP_ ` macros in ROM and to build it by `idf.py build` command , I got a lots of issues on console and it does not built.

This works for me

```c
#define MICROPY_WRAP_MP_BINARY_OP(f) IRAM_ATTR f
#define MICROPY_WRAP_MP_EXECUTE_BYTECODE(f) IRAM_ATTR f
#define MICROPY_WRAP_MP_LOAD_GLOBAL(f) IRAM_ATTR f
#define MICROPY_WRAP_MP_LOAD_NAME(f) IRAM_ATTR f
#define MICROPY_WRAP_MP_MAP_LOOKUP(f) IRAM_ATTR f
// #define MICROPY_WRAP_MP_OBJ_GET_TYPE(f) IRAM_ATTR f
// #define MICROPY_WRAP_MP_SCHED_EXCEPTION(f) IRAM_ATTR f
// #define MICROPY_WRAP_MP_SCHED_KEYBOARD_INTERRUPT(f) IRAM_ATTR f
```

> What do you mean "thread function"? I use built-in thread, it is import _thread

You said "My thread function does not work right" in the original post, so I was asking what you meant by that.

What is WavP? Does it use `@micropython.native` ?
